### PR TITLE
convert flame_wave to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
test: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-11-24-003/flame_wave.html

I still want to directly diff the initial models that are generated

**Comment by zingale** [comment]
Here's the diff of plt00000:
```
            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                            5.647540092e-06           1.807368569e-13
 xmom                                             0                         0
 ymom                                             0                         0
 zmom                                             0                         0
 rho_E                              2.230161768e+12           2.676523822e-13
 rho_e                              2.230161768e+12           2.676523822e-13
 Temp                                0.001085013151           9.146344741e-13
 rho_He4                            8.149072528e-08           4.809933619e-14
 rho_C12                            5.650861723e-16           1.808431584e-13
 rho_O16                            5.650861723e-16           1.808431584e-13
 rho_Ne20                           5.650861723e-16           1.808431584e-13
 rho_Mg24                           5.650861723e-16           1.808431584e-13
 rho_Si28                           5.650861723e-16           1.808431584e-13
 rho_S32                            5.650861723e-16           1.808431584e-13
 rho_Ar36                           5.650861723e-16           1.808431584e-13
 rho_Ca40                           5.650861723e-16           1.808431584e-13
 rho_Ti44                           5.650861723e-16           1.808431584e-13
 rho_Cr48                           5.650861723e-16           1.808431584e-13
 rho_Fe52                           5.650861723e-16           1.808431584e-13
 rho_Ni56                           5.647540092e-06           1.807368571e-13
 phiGrav                                          0                         0
 grav_x                                           0                         0
 grav_y                                           0                         0
 grav_z                                           0                         0
 phiRot                                           0                         0
 rho_omegadot_He4                                 0                         0
 rho_omegadot_C12                                 0                         0
 rho_omegadot_O16                                 0                         0
 rho_omegadot_Ne20                                0                         0
 rho_omegadot_Mg24                                0                         0
 rho_omegadot_Si28                                0                         0
 rho_omegadot_S32                                 0                         0
 rho_omegadot_Ar36                                0                         0
 rho_omegadot_Ca40                                0                         0
 rho_omegadot_Ti44                                0                         0
 rho_omegadot_Cr48                                0                         0
 rho_omegadot_Fe52                                0                         0
 rho_omegadot_Ni56                                0                         0
 rho_enuc                                         0                         0
 burn_weights                                     0                         0
 pressure                            1.57786361e+12           3.844740224e-13
 kineng                                           0                         0
 soundspeed                         0.0001180768013           2.747200606e-13
 Gamma_1                            7.052136652e-13           4.231882361e-13
 MachNumber                                       0                         0
 entropy                            0.0001647770405           3.107517502e-15
 thermal_cond                              37158912           1.480375331e-13
 diff_coeff                         2.045184374e-06           1.223238256e-13
 diff_term                           2.13674623e+11           2.120765288e-13
 magvort                                          0                         0
 divu                                             0                         0
 eint_E                                       31472           1.013104955e-13
 eint_e                                       31456           1.012589905e-13
 logden                             1.763922342e-12           2.353524279e-13
 StateErr_0                         5.647540092e-06           1.807368569e-13
 StateErr_1                          0.001085013151           9.146344741e-13
 StateErr_2                         4.218847494e-15           4.218847499e-15
 X(He4)                             4.218847494e-15           4.218847499e-15
 X(C12)                             1.163222736e-25           1.163222738e-15
 X(O16)                             1.163222736e-25           1.163222738e-15
 X(Ne20)                            1.163222736e-25           1.163222738e-15
 X(Mg24)                            1.163222736e-25           1.163222738e-15
 X(Si28)                            1.163222736e-25           1.163222738e-15
 X(S32)                             1.163222736e-25           1.163222738e-15
 X(Ar36)                            1.163222736e-25           1.163222738e-15
 X(Ca40)                            1.163222736e-25           1.163222738e-15
 X(Ti44)                            1.163222736e-25           1.163222738e-15
 X(Cr48)                            1.163222736e-25           1.163222738e-15
 X(Fe52)                            1.163222736e-25           1.163222738e-15
 X(Ni56)                            3.996802889e-15           3.996802893e-15
 abar                               4.263256415e-14           7.612957903e-16
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 t_sound_t_enuc                                   0                         0
 enuc                                             0                         0
 magvel                                           0                         0
 radvel                                           0                         0
 circvel                                          0                         0
 magmom                                           0                         0
 angular_momentum_x                               0                         0
 angular_momentum_y                               0                         0
 angular_momentum_z                               0                         0
 maggrav                                          0                         0
 level = 1
 density                            6.854534149e-06           1.996171264e-13
 xmom                                             0                         0
 ymom                                             0                         0
 zmom                                             0                         0
 rho_E                              2.550136832e+12           2.664261423e-13
 rho_e                              2.550136832e+12           2.664261423e-13
 Temp                                0.003656208515           2.683646268e-12
 rho_He4                            1.147855073e-07           4.562532078e-14
 rho_C12                            6.856494539e-16            1.99674217e-13
 rho_O16                            6.856494539e-16            1.99674217e-13
 rho_Ne20                           6.856494539e-16            1.99674217e-13
 rho_Mg24                           6.856494539e-16            1.99674217e-13
 rho_Si28                           6.856494539e-16            1.99674217e-13
 rho_S32                            6.856494539e-16            1.99674217e-13
 rho_Ar36                           6.856494539e-16            1.99674217e-13
 rho_Ca40                           6.856494539e-16            1.99674217e-13
 rho_Ti44                           6.856494539e-16            1.99674217e-13
 rho_Cr48                           6.856494539e-16            1.99674217e-13
 rho_Fe52                           6.856494539e-16            1.99674217e-13
 rho_Ni56                           6.854534149e-06           1.996171267e-13
 phiGrav                                          0                         0
 grav_x                                           0                         0
 grav_y                                           0                         0
 grav_z                                           0                         0
 phiRot                                           0                         0
 rho_omegadot_He4                                 0                         0
 rho_omegadot_C12                                 0                         0
 rho_omegadot_O16                                 0                         0
 rho_omegadot_Ne20                                0                         0
 rho_omegadot_Mg24                                0                         0
 rho_omegadot_Si28                                0                         0
 rho_omegadot_S32                                 0                         0
 rho_omegadot_Ar36                                0                         0
 rho_omegadot_Ca40                                0                         0
 rho_omegadot_Ti44                                0                         0
 rho_omegadot_Cr48                                0                         0
 rho_omegadot_Fe52                                0                         0
 rho_omegadot_Ni56                                0                         0
 rho_enuc                                         0                         0
 burn_weights                                     0                         0
 pressure                           1.220844454e+12           2.610142011e-13
 kineng                                           0                         0
 soundspeed                         0.0001567602158           3.585649154e-13
 Gamma_1                            6.679101716e-13            4.00051325e-13
 MachNumber                                       0                         0
 entropy                            0.0006271600723           1.182756345e-14
 thermal_cond                              40517632           1.396351678e-13
 diff_coeff                         5.308538675e-06            3.17221644e-13
 diff_term                          5.599563612e+12           4.651915624e-13
 magvort                                          0                         0
 divu                                             0                         0
 eint_E                                      159504           5.134541586e-13
 eint_e                                      159504           5.134541586e-13
 logden                             1.763922342e-12            2.34072957e-13
 StateErr_0                         6.854534149e-06           1.996171264e-13
 StateErr_1                          0.003656208515           2.683646268e-12
 StateErr_2                         2.220446049e-16           2.220446052e-16
 X(He4)                             2.220446049e-16           2.220446052e-16
 X(C12)                             2.584939414e-26           2.584939417e-16
 X(O16)                             2.584939414e-26           2.584939417e-16
 X(Ne20)                            2.584939414e-26           2.584939417e-16
 X(Mg24)                            2.584939414e-26           2.584939417e-16
 X(Si28)                            2.584939414e-26           2.584939417e-16
 X(S32)                             2.584939414e-26           2.584939417e-16
 X(Ar36)                            2.584939414e-26           2.584939417e-16
 X(Ca40)                            2.584939414e-26           2.584939417e-16
 X(Ti44)                            2.584939414e-26           2.584939417e-16
 X(Cr48)                            2.584939414e-26           2.584939417e-16
 X(Fe52)                            2.584939414e-26           2.584939417e-16
 X(Ni56)                            1.110223025e-16           1.110223026e-16
 abar                               7.105427358e-15           1.268826317e-16
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 t_sound_t_enuc                                   0                         0
 enuc                                             0                         0
 magvel                                           0                         0
 radvel                                           0                         0
 circvel                                          0                         0
 magmom                                           0                         0
 angular_momentum_x                               0                         0
 angular_momentum_y                               0                         0
 angular_momentum_z                               0                         0
 maggrav                                          0                         0
```

**Comment by zingale** [comment]
Here's the diffs for a mixed H/He inputs at plt00000 (to show that the species parsing is working):
```
fcompare.gnu.ex flame_wave_H_He_500Hz_plt00000 flame_wave_H_He_500Hz_plt00000.dev/

            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                            6.929039955e-07           3.084662947e-14
 xmom                                             0                         0
 ymom                                             0                         0
 zmom                                             0                         0
 rho_E                              2.845415834e+11           5.601839815e-14
 rho_e                              2.845415834e+11           5.601839815e-14
 Temp                               0.0002170056105           1.612025422e-13
 rho_C12                             6.91720986e-17            3.07939644e-14
 rho_O14                            1.157332008e-10           8.675809331e-14
 rho_O15                            2.148681233e-10           8.673187091e-14
 rho_O16                             6.91720986e-17            3.07939644e-14
 rho_F17                             6.91720986e-17            3.07939644e-14
 rho_Mg22                            6.91720986e-17            3.07939644e-14
 rho_S30                             6.91720986e-17            3.07939644e-14
 rho_Ni56                           6.929039955e-07            3.08466295e-14
 rho_He4                            8.250935934e-09           8.659310036e-14
 rho_H1                             2.441811375e-08           8.657658994e-14
 phiGrav                                          0                         0
 grav_x                                           0                         0
 grav_y                                           0                         0
 grav_z                                           0                         0
 phiRot                                           0                         0
 rho_omegadot_C12                                 0                         0
 rho_omegadot_O14                                 0                         0
 rho_omegadot_O15                                 0                         0
 rho_omegadot_O16                                 0                         0
 rho_omegadot_F17                                 0                         0
 rho_omegadot_Mg22                                0                         0
 rho_omegadot_S30                                 0                         0
 rho_omegadot_Ni56                                0                         0
 rho_omegadot_He4                                 0                         0
 rho_omegadot_H1                                  0                         0
 rho_enuc                                         0                         0
 burn_weights                                     0                         0
 pressure                           2.966211789e+11           1.152116216e-13
 kineng                                           0                         0
 soundspeed                         0.0001624822617           2.560611258e-13
 Gamma_1                            6.943334796e-13           4.177553626e-13
 MachNumber                                       0                         0
 entropy                            0.0006523132324           1.191307311e-14
 thermal_cond                              37625856           4.178306724e-13
 diff_coeff                         6.410991773e-07           4.360826431e-13
 diff_term                           9.63117056e+10           2.699349314e-13
 magvort                                          0                         0
 divu                                             0                         0
 eint_E                                      204448           3.354743067e-13
 eint_e                                      204384           3.353692905e-13
 logden                             5.870859354e-13           7.985971607e-14
 StateErr_0                         6.929039955e-07           3.084662947e-14
 StateErr_1                         0.0002170056105           1.612025422e-13
 StateErr_2                         1.033975766e-25           1.033975766e-15
 X(C12)                             1.033975766e-25           1.033975766e-15
 X(O14)                             3.859759734e-17           1.102788496e-14
 X(O15)                             7.025630078e-17           1.080866166e-14
 X(O16)                             1.033975766e-25           1.033975766e-15
 X(F17)                             1.033975766e-25           1.033975766e-15
 X(Mg22)                            1.033975766e-25           1.033975766e-15
 X(S30)                             1.033975766e-25           1.033975766e-15
 X(Ni56)                            1.043609643e-14           1.043609644e-14
 X(He4)                              2.72004641e-15           1.088018565e-14
 X(H1)                               8.10462808e-15           1.095220011e-14
 abar                               2.273736754e-13           4.060244239e-15
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 t_sound_t_enuc                                   0                         0
 enuc                                             0                         0
 magvel                                           0                         0
 radvel                                           0                         0
 circvel                                          0                         0
 magmom                                           0                         0
 angular_momentum_x                               0                         0
 angular_momentum_y                               0                         0
 angular_momentum_z                               0                         0
 maggrav                                          0                         0
 level = 1
 density                            2.153217793e-06           8.842219966e-14
 xmom                                             0                         0
 ymom                                             0                         0
 zmom                                             0                         0
 rho_E                              8.997956485e+11            1.57112387e-13
 rho_e                              8.997956485e+11            1.57112387e-13
 Temp                                0.001060217619           7.656706904e-13
 rho_C12                             2.15105711e-16           8.833347101e-14
 rho_O14                            1.249418347e-10           7.905421536e-14
 rho_O15                            2.321485226e-10           7.909294847e-14
 rho_O16                             2.15105711e-16           8.833347101e-14
 rho_F17                             2.15105711e-16           8.833347101e-14
 rho_Mg22                            2.15105711e-16           8.833347101e-14
 rho_S30                             2.15105711e-16           8.833347101e-14
 rho_Ni56                           2.153217793e-06           8.842219974e-14
 rho_He4                            8.920324035e-09           7.901796125e-14
 rho_H1                             2.642627805e-08           7.908415504e-14
 phiGrav                                          0                         0
 grav_x                                           0                         0
 grav_y                                           0                         0
 grav_z                                           0                         0
 phiRot                                           0                         0
 rho_omegadot_C12                                 0                         0
 rho_omegadot_O14                                 0                         0
 rho_omegadot_O15                                 0                         0
 rho_omegadot_O16                                 0                         0
 rho_omegadot_F17                                 0                         0
 rho_omegadot_Mg22                                0                         0
 rho_omegadot_S30                                 0                         0
 rho_omegadot_Ni56                                0                         0
 rho_omegadot_He4                                 0                         0
 rho_omegadot_H1                                  0                         0
 rho_enuc                                         0                         0
 burn_weights                                     0                         0
 pressure                           6.941740892e+11           2.407887952e-13
 kineng                                           0                         0
 soundspeed                         0.0003197789192           4.965655831e-13
 Gamma_1                            7.625011733e-13           4.576200768e-13
 MachNumber                                       0                         0
 entropy                              0.00163936615           2.993943373e-14
 thermal_cond                              44253184           4.814190724e-13
 diff_coeff                         6.923364708e-07            4.63892574e-13
 diff_term                          2.066785239e+12           2.093139284e-13
 magvort                                          0                         0
 divu                                             0                         0
 eint_E                                      622912           9.658603491e-13
 eint_e                                      622848           9.657611135e-13
 logden                               1.5063506e-12            2.03932196e-13
 StateErr_0                         2.153217793e-06           8.842219966e-14
 StateErr_1                          0.001060217619           7.656706904e-13
 StateErr_2                          9.04728795e-26           9.047287955e-16
 X(C12)                              9.04728795e-26           9.047287955e-16
 X(O14)                             8.998878032e-18           2.571108011e-15
 X(O15)                             1.691355389e-17           2.602085216e-15
 X(O16)                              9.04728795e-26           9.047287955e-16
 X(F17)                              9.04728795e-26           9.047287955e-16
 X(Mg22)                             9.04728795e-26           9.047287955e-16
 X(S30)                              9.04728795e-26           9.047287955e-16
 X(Ni56)                            2.775557562e-15           2.775557564e-15
 X(He4)                              6.52256027e-16           2.609024109e-15
 X(H1)                              1.915134717e-15            2.58801989e-15
 abar                               2.842170943e-13           5.075305298e-15
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 t_sound_t_enuc                                   0                         0
 enuc                                             0                         0
 magvel                                           0                         0
 radvel                                           0                         0
 circvel                                          0                         0
 magmom                                           0                         0
 angular_momentum_x                               0                         0
 angular_momentum_y                               0                         0
 angular_momentum_z                               0                         0
 maggrav                                          0                         0
 level = 2
 density                            2.630054951e-06           1.065716942e-13
 xmom                                             0                         0
 ymom                                             0                         0
 zmom                                             0                         0
 rho_E                              1.460288881e+12           2.499364283e-13
 rho_e                              1.460288881e+12           2.499364283e-13
 Temp                                0.004473939538           3.206641275e-12
 rho_C12                            2.628106066e-16           1.064927241e-13
 rho_O14                            1.364242053e-10           8.439853566e-14
 rho_O15                            2.532942744e-10           8.437689505e-14
 rho_O16                            2.628106066e-16           1.064927241e-13
 rho_F17                            2.628106066e-16           1.064927241e-13
 rho_Mg22                           2.628106066e-16           1.064927241e-13
 rho_S30                            2.628106066e-16           1.064927241e-13
 rho_Ni56                           2.630054951e-06           1.065716943e-13
 rho_He4                            9.735231288e-09           8.431751316e-14
 rho_H1                             2.881279215e-08            8.43072941e-14
 phiGrav                                          0                         0
 grav_x                                           0                         0
 grav_y                                           0                         0
 grav_z                                           0                         0
 phiRot                                           0                         0
 rho_omegadot_C12                                 0                         0
 rho_omegadot_O14                                 0                         0
 rho_omegadot_O15                                 0                         0
 rho_omegadot_O16                                 0                         0
 rho_omegadot_F17                                 0                         0
 rho_omegadot_Mg22                                0                         0
 rho_omegadot_S30                                 0                         0
 rho_omegadot_Ni56                                0                         0
 rho_omegadot_He4                                 0                         0
 rho_omegadot_H1                                  0                         0
 rho_enuc                                         0                         0
 burn_weights                                     0                         0
 pressure                           5.921686159e+11           2.015765482e-13
 kineng                                           0                         0
 soundspeed                         0.0003955960274           6.117162201e-13
 Gamma_1                            5.180300633e-13           3.107674757e-13
 MachNumber                                       0                         0
 entropy                             0.001724720001           3.149823496e-14
 thermal_cond                              47448064           5.094053852e-13
 diff_coeff                         7.466878742e-07            4.98657348e-13
 diff_term                          2.499527311e+13           1.787948864e-12
 magvort                                          0                         0
 divu                                             0                         0
 eint_E                                      656832           9.899980198e-13
 eint_e                                      656800           9.899497884e-13
 logden                             1.505906511e-12           2.037122188e-13
 StateErr_0                         2.630054951e-06           1.065716942e-13
 StateErr_1                          0.004473939538           3.206641275e-12
 StateErr_2                         6.462348536e-26           6.462348539e-16
 X(C12)                             6.462348536e-26           6.462348539e-16
 X(O14)                             2.168404345e-18           6.195440989e-16
 X(O15)                             3.469446952e-18           5.337610699e-16
 X(O16)                             6.462348536e-26           6.462348539e-16
 X(F17)                             6.462348536e-26           6.462348539e-16
 X(Mg22)                            6.462348536e-26           6.462348539e-16
 X(S30)                             6.462348536e-26           6.462348539e-16
 X(Ni56)                            1.110223025e-16           1.110223026e-16
 X(He4)                             1.387778781e-16           5.551115126e-16
 X(H1)                              2.220446049e-16           3.000602771e-16
 abar                               7.105427358e-15           1.268826325e-16
 x_velocity                                       0                         0
 y_velocity                                       0                         0
 z_velocity                                       0                         0
 t_sound_t_enuc                                   0                         0
 enuc                                             0                         0
 magvel                                           0                         0
 radvel                                           0                         0
 circvel                                          0                         0
 magmom                                           0                         0
 angular_momentum_x                               0                         0
 angular_momentum_y                               0                         0
 angular_momentum_z                               0                         0
 maggrav                                          0                         0
```

File: Exec/science/flame_wave/problem_setup.H [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+    Real x = problo[0] + dx[0] * (static_cast<Real>(i) + 0.5_rt) - problem::center[0];
+
+    Real y = 0.0;
+#if AMREX_SPACEDIM >= 2
+    y = problo[1] + dx[1] * (static_cast<Real>(j) + 0.5_rt) - problem::center[1];
+#endif
+
+    Real z = 0.0;
+#if AMREX_SPACEDIM == 3
+    z = problo[2] + dx[2] * (static_cast<Real>(k) + 0.5_rt);
```
Comment (maxpkatz) [inline Exec/science/flame_wave/problem_setup.H:233]: What does it mean that `x` and `y` subtract off center but `z` does not?

Comment (zingale) [reply Exec/science/flame_wave/problem_setup.H:233]: in 2-d, the center is 0, so this is a no op.  In 3-d, we want to be in the geometric center laterally, so we subtract off the center, but for z, we want to know the height from the bottom of the domain for interpolating, so we don't subtract off center.

I suppose I could just set the center in z to be 0 to be consistent.

**Review by maxpkatz — APPROVED** [review]

**Pull request merged**